snarky_universe/snarky_universe.ml

```diff
@@ -0,0 +1,271 @@
+module Cond (Impl : sig
+  module Boolean : sig
+    type var
+  end
+end) (M : sig
+  type t
+
+  val if_ : Impl.Boolean.var -> then_:t -> else_:t -> t
+end) : Intf.Cond_intf with type bool := Impl.Boolean.var and type t := M.t =
+struct
+  open Impl
+  open M
+
+  type cond = Boolean.var * t
+
+  let ( -? ) c x : cond = (c, x)
+
+  let ( -: ) (c, x) y = if_ c ~then_:x ~else_:y
+end
+
+let impl_of_curve = function
+  | Curve.Bn128 ->
+      let module Backend = Snarky.Backends.Bn128.Default in
+      let module Run =
+        Snarky.Snark.Run.Make
+          (Backend)
+          (struct
+            type t = unit
+          end)
+      in
+      (module Run : Snarky.Snark_intf.Run with type prover_state = unit)
+
+module Make (C : sig
+  val curve : Curve.t
+end)
+() : Intf.S = struct
+  module Impl = (val impl_of_curve C.curve)
```

Comment:
I think `struct .. end` is needed here.

Reply:
functors can take a "unit" argument like this to make them generative rather than applicative. Basically applicative means you get type equalities between different applications whereas generative means you don't.